File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.cpp
class Solution {
public:
vector<long long>dp;
int dfs(int node,int parent,vector<int>adj[],vector<int>&values,int k,int&count){
  if(dp[node]!=-1){
    return dp[node];
  }
long long   sum=0;
    for(auto it:adj[node])
    {
        if(it!=parent){
             
            sum+=dfs(it,node,adj,values,k,count);
            sum=sum%k;
           

        }
    }
    sum+=values[node];
    sum=sum%k;
    if(sum==0){
        count++;
    }
    return dp[node]=sum;



}
    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) {
        if(n==0){
            return 0;
        }
        vector<int>vis(n,0);
        
       vector<int> adj[n];
        for(auto it:edges){
            adj[it[0]].push_back(it[1]);
            adj[it[1]].push_back(it[0]);

        }
      
      int count=0;
      dp.resize(n,-1);
   
           dfs(0,-1,adj,values,k,count);
           return count;

        

    }
};

File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.java
class Solution {
   public int count=0;
    public  long dfs(int i,ArrayList<Integer>[]adj,int parent,int k,int []values,HashMap<String, Long> dp){
        String key1=i+","+parent;
        if(dp.getOrDefault(key1, (long)-1)!=-1)return dp.get(key1);
        long sum=0;
        for(int ele:adj[i]){
            if(ele!=parent){
               sum+=dfs(ele,adj,i,k,values,dp);
               sum=sum%k;

            }
        }
        sum+=values[i];
        sum=sum%k;
        if(sum==0){
            count++;
        }
        String key=i+","+parent;
       dp.put(key,sum);
       return sum;

    }
    public int maxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        ArrayList<Integer>[]adj=new ArrayList[n];
        for(int i=0;i<n;i++){
            adj[i]=new ArrayList<>();
        }
        HashMap<String, Long> dp = new HashMap<>();
        for(int i=0;i<edges.length;i++){
            adj[edges[i][0]].add(edges[i][1]);
            adj[edges[i][1]].add(edges[i][0]);
        }
        int vis[]=new int[n];
        dfs(0,adj,-1,k,values,dp);
        return count;

    }
}